source/common/quic/envoy_quic_client_packet_writer_factory.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+#include "envoy/network/address.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/server/factory_context.h"
+
+#include "source/common/quic/envoy_quic_packet_writer.h"
+
+#include "quiche/quic/core/quic_path_validator.h"
+
+namespace Envoy {
+namespace Quic {
+
+// An extension interface to creating customized UDP sockets and creating
+// QUIC packet writers for upstream connections based on such sockets.
+class QuicClientPacketWriterFactory {
+public:
+  virtual ~QuicClientPacketWriterFactory() = default;
+
+  struct CreationResult {
+    // Not null.
+    std::unique_ptr<EnvoyQuicPacketWriter> writer_;
```

Comment:
I'm not a big fan of the using declarations with EnvoyQuicPacketWriterPtr, but up to you if you want to do that here for consistency